File: dsa/Tree/HeightOfBinaryTree.java
import java.util.LinkedList;
import java.util.Queue;

/**
 * HeightOfBinaryTree
 */

class Node {
    int val;
    Node left, right;

    Node(int val) {
        this.val = val;
        left = null;
        right = null;
    }
}

class Solution {
    int maxDepthDFS(Node root) {
        if (root == null)
            return 0;
        int lh = maxDepthDFS(root.left);
        int rh = maxDepthDFS(root.right);

        return 1 + Math.max(lh, rh);
    }

    int maxDepthBFS(Node rootNode) {
        if (rootNode == null)
            return 0;

        Queue<Node> queue = new LinkedList<>();
        queue.add(rootNode);

        int level = 0;
        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                Node node = queue.poll();
                if (node.left != null)
                    queue.add(node.left);
                if (node.right != null)
                    queue.add(node.right);
            }
            level++;
        }
        return level;
    }
}

public class HeightOfBinaryTree {
    public static void main(String[] args) {
        Node root = new Node(1);
        root.left = new Node(2);
        root.right = new Node(3);
        root.left.left = new Node(4);
        root.left.right = new Node(5);
        root.left.right.right = new Node(6);
        root.left.right.right.right = new Node(7);

        Solution solution = new Solution();
        int depth = solution.maxDepthDFS(root);
        int depth2 = solution.maxDepthBFS(root);

        System.out.println(depth);
        System.out.println(depth2);
    }
}

File: dsa/Tree/HeightOfBinaryTree.kt
import java.util.LinkedList
import java.util.Queue

class Node(val `val`: Int) {
    var left: Node? = null
    var right: Node? = null
}

class Solution {
    fun maxDepthDFS(root: Node?): Int {
        if (root == null) return 0
        val leftDepth = maxDepthDFS(root.left)
        val rightDepth = maxDepthDFS(root.right)
        return 1 + maxOf(leftDepth, rightDepth)
    }

    fun maxDepthBFS(root: Node?): Int {
        if (root == null) return 0

        val queue: Queue<Node> = LinkedList()
        queue.add(root)

        var level = 0
        while (queue.isNotEmpty()) {
            val size = queue.size
            for (i in 0 until size) {
                val node = queue.poll()
                node.left?.let { queue.add(it) }
                node.right?.let { queue.add(it) }
            }
            level++
        }
        return level
    }
}

fun main() {
    val root = Node(1).apply {
        left = Node(2).apply {
            left = Node(4)
            right = Node(5).apply {
                right = Node(6).apply {
                    right = Node(7)
                }
            }
        }
        right = Node(3)
    }

    val solution = Solution()
    val depthDFS = solution.maxDepthDFS(root)
    val depthBFS = solution.maxDepthBFS(root)

    println("Depth calculated using DFS: $depthDFS")
    println("Depth calculated using BFS: $depthBFS")
}
